cgi.go

```diff
@@ -8,214 +9,155 @@ import (
-				fc.Env["SERVER_PORT"] = "443"
-			} else {
-				fc.Env["SERVER_PORT"] = "80"
-			}
+		switch rs {
+		case "https":
+			reqPort = "443"
+		case "http":
+			reqPort = "80"
 		}
 	}
 
+	allocServerVariable(&cArr, fc.env, serverName, "SERVER_NAME", reqHost)
+	if reqPort != "" {
+		allocServerVariable(&cArr, fc.env, serverPort, "SERVER_PORT", reqPort)
+	}
+
 	// Variables defined in CGI 1.1 spec
 	// Some variables are unused but cleared explicitly to prevent
 	// the parent environment from interfering.
-	// We never override an entry previously set
-	if _, ok := fc.Env["REMOTE_IDENT"]; !ok {
-		fc.Env["REMOTE_IDENT"] = "" // Not used
-	}
-	if _, ok := fc.Env["AUTH_TYPE"]; !ok {
-		fc.Env["AUTH_TYPE"] = "" // Not used
-	}
-	if _, ok := fc.Env["CONTENT_LENGTH"]; !ok {
-		fc.Env["CONTENT_LENGTH"] = request.Header.Get("Content-Length")
-	}
-	if _, ok := fc.Env["CONTENT_TYPE"]; !ok {
-		fc.Env["CONTENT_TYPE"] = request.Header.Get("Content-Type")
-	}
-	if _, ok := fc.Env["GATEWAY_INTERFACE"]; !ok {
-		fc.Env["GATEWAY_INTERFACE"] = "CGI/1.1"
-	}
-	if _, ok := fc.Env["QUERY_STRING"]; !ok {
-		fc.Env["QUERY_STRING"] = request.URL.RawQuery
-	}
-	if _, ok := fc.Env["REQUEST_METHOD"]; !ok {
-		fc.Env["REQUEST_METHOD"] = request.Method
-	}
-	if _, ok := fc.Env["SERVER_PROTOCOL"]; !ok {
-		fc.Env["SERVER_PROTOCOL"] = request.Proto
-	}
-	if _, ok := fc.Env["SERVER_SOFTWARE"]; !ok {
-		fc.Env["SERVER_SOFTWARE"] = "FrankenPHP"
-	}
-	if _, ok := fc.Env["HTTP_HOST"]; !ok {
-		fc.Env["HTTP_HOST"] = request.Host // added here, since not always part of headers
-	}
-	if _, ok := fc.Env["REQUEST_URI"]; !ok {
-		fc.Env["REQUEST_URI"] = request.URL.RequestURI()
-	}
 
-	// compliance with the CGI specification requires that
-	// PATH_TRANSLATED should only exist if PATH_INFO is defined.
-	// Info: https://www.ietf.org/rfc/rfc3875 Page 14
-	if fc.Env["PATH_INFO"] != "" {
-		fc.Env["PATH_TRANSLATED"] = sanitizedPathJoin(fc.Env["DOCUMENT_ROOT"], fc.Env["PATH_INFO"]) // Info: http://www.oreilly.com/openbook/cgi/ch02_04.html
-	}
-
-	// Add all HTTP headers to env variables
-	for field, val := range request.Header {
-		k := "HTTP_" + headerNameReplacer.Replace(strings.ToUpper(field))
-		if _, ok := fc.Env[k]; !ok {
-			fc.Env[k] = strings.Join(val, ", ")
-		}
-	}
-
-	if _, ok := fc.Env["REMOTE_USER"]; !ok {
-		var (
-			authUser string
-			ok       bool
-		)
-		authUser, fc.authPassword, ok = request.BasicAuth()
-		if ok {
-			fc.Env["REMOTE_USER"] = authUser
-		}
-	}
+	// These values can not be override
+	cArr[contentLength] = C.CString(request.Header.Get("Content-Length"))
 
-	fc.populated = true
+	allocServerVariable(&cArr, fc.env, gatewayInterface, "GATEWAY_INTERFACE", "CGI/1.1")
+	allocServerVariable(&cArr, fc.env, serverProtocol, "SERVER_PROTOCOL", request.Proto)
+	allocServerVariable(&cArr, fc.env, serverSoftware, "SERVER_SOFTWARE", "FrankenPHP")
+	allocServerVariable(&cArr, fc.env, httpHost, "HTTP_HOST", request.Host) // added here, since not always part of headers
```

Comment:
Does `request.Host` come from the headers, if it is set there? I know many WordPress installs typically set the site name from this variable.

Reply:
More or less yes https://pkg.go.dev/net/http#Request.Host